engine/crates/operation-checks/src/aggregate_field_usage.rs

```diff
@@ -0,0 +1,461 @@
+mod async_graphql;
+
+use std::collections::HashMap;
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct SelectionId(usize);
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct FieldId(usize);
+
+/// Usage count of a field in a query.
+#[derive(Debug)]
+pub struct FieldUsage<'a> {
+    pub increment: u64,
+    /// field id -> usage count
+    pub count_per_field: &'a mut HashMap<FieldId, u64>,
+}
+
+impl FieldUsage<'_> {
+    pub fn with_increment(self, new_increment: u64) -> Self {
+        Self {
+            increment: new_increment,
+            count_per_field: self.count_per_field,
+        }
+    }
+
+    /// Register a field usage.
+    fn register(&mut self, field_id: FieldId) {
+        if let Some(count) = self.count_per_field.get_mut(&field_id) {
+            *count += self.increment;
+        } else {
+            self.count_per_field.insert(field_id, self.increment);
+        }
+    }
+}
+
+/// (type name, field name) -> field type
+#[derive(Debug)]
+pub struct Schema {
+    pub fields: Vec<SchemaField>,
+    pub query_type_name: String,
+    pub mutation_type_name: String,
+    pub subscription_type_name: String,
+}
+
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub struct SchemaField {
+    type_name: String,
+    field_name: String,
+    /// The type fo the field without any wrapping type (! and []).
+    field_type: String,
```

Comment:
I think we use `base_type` in the engine for this definition

Reply:
good catch, let's align the terminology

Reply:
done